autode/ext/src/idpp.cpp

```diff
@@ -0,0 +1,904 @@
+         * for a given number of atoms
+         *
+         * Arguments:
+         *
+         *   num_atoms: number of atoms in the image
+         *
+         *   k: spring constant
+         */
+        ensure(num_atoms > 0, "Number of atoms must be > 0");
+        ensure(k > 0, "Spring constant must be positive");
+        coords = arrx::zeros(n_atoms * 3);
+        grad = arrx::zeros(n_atoms * 3);
+    }
+
+    double Image::get_tau_k_fac(arrx::array1d& tau,
+                                const Image& img_m1,
+                                const Image& img_p1,
+                                const bool force_lc) const {
+        /* Get the unit tangent and the pre-factor required to calculate
+         * the spring force, from the previous and next image
+         *
+         * Arguments:
+         *
+         *   tau: (out) The array where the tangent is stored
+         *
+         *   img_m1: Previous image
+         *
+         *   img_p1: Next image
+         *
+         *   force_lc: Whether to force using linear combination tangent
+         *             even if image is not extrema in energy
+         */
+
+        auto dv_max = std::max(
+            std::abs(img_p1.en - en),
+            std::abs(img_m1.en - en)
+        );
+        auto dv_min = std::min(
+            std::abs(img_p1.en - en),
+            std::abs(img_m1.en - en)
+        );
+
+        arrx::array1d tau_p = img_p1.coords - coords;
+        arrx::array1d tau_m = coords - img_m1.coords;
+
+        double tau_p_norm = arrx::norm_l2(tau_p);
+        double tau_m_norm = arrx::norm_l2(tau_m);
+
+        ensure(tau_p_norm > 1e-10 && tau_m_norm > 1e-10,
+                "Failed to get tangent, images too close");
+
+        tau_p /= tau_p_norm;
+        tau_m /= tau_m_norm;
+
+        bool use_lc = force_lc;
+        if (force_lc) {
+
+        } else if (img_m1.en < en && en < img_p1.en) {
+            tau = tau_p;
+        } else if (img_p1.en < en && en < img_m1.en) {
+            tau = tau_m;
+        } else {
+            use_lc = true;
+        }
+
+        if (use_lc) {
+            if (img_m1.en < img_p1.en) {
+                tau = tau_p * dv_max + tau_m * dv_min;
+            } else if (img_p1.en < img_m1.en) {
+                tau = tau_m * dv_max + tau_p * dv_min;
+            } else {  // equal energies
+                tau = tau_p + tau_m;
+            }
+            double tau_norm = arrx::norm_l2(tau);
+            if (tau_norm < 1e-10) {  // sometimes dv_max, dv_min are small
+                tau = tau_p + tau_m;
+                tau_norm = arrx::norm_l2(tau);
+                ensure(tau_norm > 1e-10, "Tangent vector too small");
+            }
+            tau /= tau_norm;
+        }
+        return tau_p_norm * img_p1.k_spr - tau_m_norm * img_m1.k_spr;
+    }
+
+    void Image::update_neb_grad(const Image& img_m1,
+                                const Image& img_p1,
+                                bool force_lc) {
```

Comment:
looks like this function is only called with `force_lc  = false` – could remove the flag?